omp/base/index_set_kernels.cpp

```diff
@@ -148,29 +148,29 @@ GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_INDEX_SET_POPULATE_KERNEL);
 template <typename IndexType>
 void global_to_local(std::shared_ptr<const DefaultExecutor> exec,
                      const IndexType index_space_size,
-                     const Array<IndexType>* subset_begin,
-                     const Array<IndexType>* subset_end,
-                     const Array<IndexType>* superset_indices,
-                     const Array<IndexType>* global_indices,
-                     Array<IndexType>* local_indices, const bool is_sorted)
+                     const IndexType num_subsets, const IndexType* subset_begin,
+                     const IndexType* subset_end,
+                     const IndexType* superset_indices,
+                     const IndexType num_indices,
+                     const IndexType* global_indices, IndexType* local_indices,
+                     const bool is_sorted)
 {
 #pragma omp parallel for
-    for (size_type i = 0; i < global_indices->get_num_elems(); ++i) {
-        auto index = global_indices->get_const_data()[i];
-        GKO_ASSERT(index < index_space_size);
-        auto bucket =
-            std::distance(subset_begin->get_const_data(),
-                          std::upper_bound(subset_begin->get_const_data(),
-                                           subset_begin->get_const_data() +
-                                               subset_begin->get_num_elems(),
-                                           index));
+    for (size_type i = 0; i < num_indices; ++i) {
+        auto index = global_indices[i];
+        if (index > index_space_size) {
+            local_indices[i] = invalid_index<IndexType>();
+            continue;
+        }
+        const auto bucket = std::distance(
+            subset_begin,
+            std::upper_bound(subset_begin, subset_begin + num_subsets, index));
```

Comment:
shouldn't this be using subset_end? Then you also won't need the following line

Reply:
I think that is more or less equivalent ? I think you need to have a shifted index anyway to access the superset_indices later.

Reply:
yes, I guess so. If I read upper_bound, I am thinking: you are looking for something that is larger than the entry, and subset_end would be less surprising